crates/core/component/dex/src/component/position_manager.rs

```diff
@@ -188,38 +195,81 @@ pub trait PositionManager: StateWrite + PositionRead {
         }
 
         // Validate that the position ID doesn't collide
-        self.check_position_id_unused(&position.id()).await?;
+        if let Some(existing) = self.position_by_id(&position.id()).await? {
+            anyhow::bail!(
+                "attempted to open a position with ID {}, which already exists with state {:?}",
+                position.id(),
+                existing
+            );
+        }
 
         // Credit the DEX for the inflows from this position.
         self.vcb_credit(position.reserves_1()).await?;
         self.vcb_credit(position.reserves_2()).await?;
 
         // Finally, record the new position state.
         self.record_proto(event::position_open(&position));
-        self.update_position(position).await?;
+        self.update_position(None, position).await?;
 
         Ok(())
     }
 
     /// Record execution against an opened position.
+    ///
+    /// The `context` parameter records the global context of the path in which
+    /// the position execution happened. This may be completely different than
+    /// the trading pair of the position itself, and is used to link the
+    /// micro-scale execution (processed by this method) with the macro-scale
+    /// context (a swap or arbitrage).
     #[tracing::instrument(level = "debug", skip_all)]
-    async fn position_execution(&mut self, mut position: Position) -> Result<()> {
+    async fn position_execution(
+        &mut self,
+        mut new_state: Position,
+        context: DirectedTradingPair,
+    ) -> Result<()> {
+        let prev_state = self
+            .position_by_id(&new_state.id())
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("withdrew from unknown position {}", new_state.id()))?;
+
+        if !matches!(&prev_state.state, position::State::Opened) {
```

Comment:
I like to use `anyhow::ensure!(invariant, msg)` for guards

Reply:
Good idea, updated